crates/blockchain-tree/src/shareable.rs

```diff
@@ -83,6 +83,22 @@ where
         res
     }
 
+    fn update_block_hashes_and_clear_buffered(
+        &self,
+    ) -> RethResult<BTreeMap<BlockNumber, BlockHash>> {
+        let mut tree = self.tree.write();
+        let res = tree.update_block_hashes();
+
+        if let Ok(Some((block, _))) =
+            res.as_ref().map(|last_canonical_hashes| last_canonical_hashes.last_key_value())
+        {
+            tree.remove_old_blocks(*block);
+        }
+
+        tree.update_chains_metrics();
+        res
+    }
+
```

Comment:
we have a pattern of deferring the `ShareableBlockchainTree` method implementations fully to the underlying `self.tree`, can we do it here as well?